server/functions/0_overload_deduction.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2023 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package functions
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/types"
+)
+
+// OverloadDeduction handles resolving which function to call by iterating over the parameter expressions. This also
+// handles casting between types if an exact function match is not found.
+type OverloadDeduction struct {
+	Function      reflect.Value
+	ReturnSqlType sql.Type
+	ReturnValType ParameterType
+	Parameter     [ParameterType_Length]*OverloadDeduction
+}
+
+// Resolve returns an overload that either matches the given parameters exactly, or is a viable match after casting.
+// This will modify the parameter slice in-place. Returns a nil OverloadDeduction if a viable match is not found.
```

Comment:
Better to not modify things in place when we can avoid it. It makes things like prepared statement execution more challenging, since you can't count on being able to reuse the same query plan multiple times in all cases.

Reply:
I was confused earlier when we were speaking. This is modifying the passed-in parameter in-place, which is a function-local variable of the caller. Nothing "permanent" is actually modified.